### PR TITLE
[DistDGL][Robustness]Replacing numpy's unique with custom implementation

### PR DESCRIPTION
**Pull request by kylasa**
numpy's unique is very memory expensive. It is about 8x times the input size. For large graphs this will be an issue. Consider the case where we have 3 billion edges, in this case, this function will consume memory which is as high as 550Gb. This is limiting the size of the input graph we can partition on a single machine. 

Replacing numpy's unique with memory efficient custom implementation. which is atmost 3x memory expensive. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [X] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [X] All changes have test coverage
- [X] Code is well-documented
- [X] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by kylasa** [comment]
https://github.com/dmlc/dgl/issues/5395 is opened to update the numpy version. Once this is fixed CI tests are expected to pass.

File: tools/distpartitioning/convert_partition.py [hunk 1]
```diff
@@ -21,6 +21,94 @@
 from utils import get_idranges, memory_snapshot, read_json
 
 
+def _get_unique_invidx(srcids, dstids, uniq_nids):
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:24]: pls add dedicated testcase for this newly added function.

File: tools/distpartitioning/convert_partition.py [hunk 2]
```diff
@@ -18,7 +18,129 @@
...
+        a list of integers. These are inverse indices, which will be indices
+        from the unique elements list specifying the elements from the
+        input array, srcids
+    numpy array :
+        a list of integers. These are inverse indices, which will be indices
+        from the unique elements list specifying the elements from the
+        input array, dstids
+
+    """
+    mask = np.in1d(srcids, nids, invert=True, kind="table")
```
Comment (Rhett-Ying) [inline tools/distpartitioning/convert_partition.py:70]: `kind=` requires `numpy >=1.24` which not available in CI.

File: tests/tools/test_convert_partition.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    srcids = np.random.randint(0, num_nodes, (num_edges,))
+    dstids = np.random.randint(nid_begin, nid_end, (num_edges,))
+
+    # generate unique node-ids for any
+    # partition. This list should be sorted.
+    # This is equivilant to shuffle_nids in a partition
+    unique_nids = np.arange(nid_begin, nid_end)
+
+    # invoke the test target
+    uniques, idxes, src_ids, dst_ids = _get_unique_invidx(
```
Comment (classicsong) [inline tests/tools/test_convert_partition.py:33]: Can you compare the output with a implementation using np.unique?

File: tests/tools/test_convert_partition.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+from tools.distpartitioning import utils
+
+from tools.distpartitioning.convert_partition import _get_unique_invidx
+
+
+@pytest.mark.parametrize(
+    "num_nodes, num_edges, nid_begin, nid_end",
+    [
+        [4000, 40000, 0, 1000],
```
Comment (classicsong) [inline tests/tools/test_convert_partition.py:15]: Can you add test cases as [1,1,0,1], [4000,0,1000,4000], etc. to test the corner cases.

File: tools/distpartitioning/convert_partition.py (hunk 2 above)
Comment (kylasa) [reply tools/distpartitioning/convert_partition.py]: Testing the new functionality without kind.

File: tests/tools/test_convert_partition.py (hunk 4 above)
Comment (kylasa) [reply tests/tools/test_convert_partition.py:15]: Done.

File: tests/tools/test_convert_partition.py (hunk 3 above)
Comment (kylasa) [reply tests/tools/test_convert_partition.py:33]: I don't understand. For every call do you want me to compare the workaround solution with numpy's unique ??

File: tests/tools/test_convert_partition.py (hunk 4 above)
Comment (kylasa) [reply tests/tools/test_convert_partition.py:15]: [4000, 0, 1000, 4000] cannot be a valid test case, since we always need a non-empty edge-list to process. One of the return parameters of this function is the edge-list itself, which is over-written with the return parameters.

File: tests/tools/test_convert_partition.py [hunk 5]
```diff
@@ -4,25 +4,115 @@
...
+        [4000, 40000, 2000, 3000],
+        [4000, 40000, 3000, 4000],
+        [4000, 100, 0, 1000],
+        [4000, 100, 1000, 2000],
+        [4000, 100, 2000, 3000],
+        [4000, 100, 3000, 4000],
+        [1, 1, 0, 1],
+    ],
+)
+def test_get_unique_invidx_with_numpy(num_nodes, num_edges, nid_begin, nid_end):
```
Comment (classicsong) [inline tests/tools/test_convert_partition.py:26]: What is the difference between test_get_unique_invidx_with_numpy and test_get_unique_invidx?

File: tools/distpartitioning/convert_partition.py [hunk 6]
```diff
@@ -65,44 +78,46 @@ def _get_unique_invidx(srcids, dstids, nids):
...
-        f"It is not a super-set. "
-        f"len(dstids_int_nids) = {len(dstids_int_nids)}, len(nids) = {len(nids)}"
-    )
-
-    # validation check here.
-    assert np.all(gold_uniques.sort() == uniques.sort()), (
-        f"Our assumption about gold_unique is NOT right. "
-        f"len(gold_uniques) == {len(gold_uniques)} "
-        f"len(uniques) == {len(uniques)}"
+    # build uniques and idxes, first and second return parameters
```
Comment (classicsong) [inline tools/distpartitioning/convert_partition.py:117]: Did you change the implementation?

File: tools/distpartitioning/convert_partition.py [hunk 7]
```diff
@@ -65,44 +78,46 @@ def _get_unique_invidx(srcids, dstids, nids):
...
+    uniques = np.concatenate([uniques, nids])
+    idxes = np.concatenate(
+        [idxes, len(srcids) + len(dstids) + np.arange(len(nids))]
```
Comment (classicsong) [inline tools/distpartitioning/convert_partition.py:120]: Can you prepare a google doc describing the details of the `unique` implementation?

File: tests/tools/test_convert_partition.py (hunk 5 above)
Comment (kylasa) [reply tests/tools/test_convert_partition.py:26]: test_get_unique_invidx_with_numpy - is the one you requested. Comparing the results with numpy's unique results. and test_get_unique_invidx is the test case for testing our work-around solution (here I validate the results from the workaround solution). I like to keep them separated.

File: tools/distpartitioning/convert_partition.py (hunk 6 above)
Comment (kylasa) [reply tools/distpartitioning/convert_partition.py:117]: In a way yes. This newer implementation is optimized for wall-clock time. In the earlier version, which was used for benchmarking purposes on felarge and webgraphs, we used to compute the list of uniques in a step by step manner - and validating them as well. Now I removed the validation part, since the benchmarking results were highly reduced when compared to the numpy's unique implementation.

File: tools/distpartitioning/convert_partition.py (hunk 7 above)
Comment (kylasa) [reply tools/distpartitioning/convert_partition.py:120]: The doc string and the comments in the code are very extensive. If you follow these comments, you should be able understand the implementation fairly easily. But if you prefer a quip document, I will put the same text in the quip document and share it with a larger group.

Comment (kylasa) [reply tools/distpartitioning/convert_partition.py:120]: https://quip-amazon.com/x5KSAK0FamJE/Dist-Graph-Partitioning-Pipeline-Numpy-Unique-Issues

**Review by classicsong — APPROVED** [review]

**Comment by zheng-da** [comment]
The test script doesn't run in CI?
